[FSDP] Warning when fail to clone

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74946

Warn instead of hard failure when fail to clone state_dict, as this
param might not be managed by FSDP and thus we do not expect to clone it.

Differential Revision: [D35242306](https://our.internmc.facebook.com/intern/diff/D35242306/)

rohan-varma committed:
[FSDP] Warning when fail to clone

Warn instead of hard failure when fail to clone state_dict, as this
param might not be managed by FSDP and thus we do not expect to clone it.

Differential Revision: [D35242306](https://our.internmc.facebook.com/intern/diff/D35242306/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -886,8 +886,17 @@ def _full_post_state_dict_hook(
+                    state_dict[key]._has_been_cloned = True  # type: ignore[attr-defined]
+                except BaseException as e:
+                    warnings.warn(
+                        f"Failed to clone() tensor with name {key}. This may mean"
+                        "that this state_dict entry could point to invalid memory"
+                        "regions after returning from state_dict() call if this"
+                        "parameter is managed by FSDP. Please check clone"
+                        f"implementation of {key}. Error: {str(e)}"
```
nit: space after "mean", "memory", etc.

rohan-varma committed:
Update on "[FSDP] Warning when fail to clone"


Warn instead of hard failure when fail to clone state_dict, as this
param might not be managed by FSDP and thus we do not expect to clone it.

Differential Revision: [D35242306](https://our.internmc.facebook.com/intern/diff/D35242306/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Warning when fail to clone"


Warn instead of hard failure when fail to clone state_dict, as this
param might not be managed by FSDP and thus we do not expect to clone it.

Differential Revision: [D35242306](https://our.internmc.facebook.com/intern/diff/D35242306/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP] Warning when fail to clone"


Warn instead of hard failure when fail to clone state_dict, as this
param might not be managed by FSDP and thus we do not expect to clone it.

Differential Revision: [D35242306](https://our.internmc.facebook.com/intern/diff/D35242306/)

[ghstack-poisoned]

osalpekar commented:
Can we merge this with GHF when it's ready?
